File: packages/typescript/goldenpipe/src/core/adapters/flow.ts
/**
 * GoldenFlow adapter — wraps GoldenFlow-JS `TransformEngine.transformDf`.
 * Port of goldenpipe/adapters/flow.py.
 *
 * Shape note: GoldenFlow-JS exposes `new TransformEngine(config).transformDf(rows)`
 * which returns `{ rows, columns, manifest }` (the Python sibling's
 * `transform_df(df)` returns an object with `.df` + `.manifest`). We read
 * `.rows` back into `ctx.df` and surface `.manifest` as an artifact.
 *
 * Edge-safe: no `node:` imports (GoldenFlow-JS core is edge-safe).
 */

import { TransformEngine } from "goldenflow/core";
import type { GoldenFlowConfig, TransformSpec } from "goldenflow/core";
import type { PipeContext, Stage, StageResult, Row } from "../models.js";
import { StageStatus } from "../models.js";
import { enrichContextsFromFlow, type ColumnContext, type ManifestRecordLike } from "../columnContext.js";
import { repairTransformSpecs, mergeTransforms } from "../repairHost.js";

export const TransformStage: Stage = {
  info: { name: "goldenflow.transform", produces: ["df", "manifest"], consumes: ["df"] },

  validate(ctx: PipeContext): void {
    if (ctx.df === null) {
      throw new Error("TransformStage: no df in context");
    }
  },

  async run(ctx: PipeContext): Promise<StageResult> {
    const rows = (ctx.df ?? []) as Row[];
    const rawCfg: Record<string, unknown> = { ...(ctx.stageConfig ?? {}) };
    const apply = rawCfg["apply_repairs"] === true;
    delete rawCfg["apply_repairs"];

    let config: Partial<GoldenFlowConfig> | undefined;
    if (apply) {
      const plan = ctx.artifacts["repair_plan"] as Parameters<typeof repairTransformSpecs>[0];
      const { specs, skipped } = repairTransformSpecs(plan);
      const userTransforms = (rawCfg["transforms"] as TransformSpec[] | undefined) ?? [];
      if (skipped.length > 0) {
        ctx.reasoning["repair_skipped"] = skipped.map((s) => `${s.column}:${s.op}`).join("; ");
      }
      if (specs.length > 0 || userTransforms.length > 0) {
        config = { ...rawCfg, transforms: mergeTransforms(userTransforms, specs) } as Partial<GoldenFlowConfig>;
      } else {
        config = Object.keys(rawCfg).length > 0 ? (rawCfg as Partial<GoldenFlowConfig>) : undefined;
      }
    } else {
      config = Object.keys(rawCfg).length > 0 ? (rawCfg as Partial<GoldenFlowConfig>) : undefined;
    }

    const engine = new TransformEngine(config);
    const result = engine.transformDf(rows);

    ctx.df = [...result.rows] as Row[];
    ctx.artifacts["manifest"] = result.manifest;

    // Enrich column contexts with transform information (best-effort).
    const contexts = ctx.artifacts["column_contexts"];
    if (Array.isArray(contexts)) {
      try {
        const records: ManifestRecordLike[] = result.manifest.records.map((r) => ({
          column: r.column,
          transform: r.transform,
          affectedRows: r.affectedRows,
        }));
        enrichContextsFromFlow(contexts as ColumnContext[], records);
      } catch {
        /* best-effort: never break the pipeline on enrichment failure */
      }
    }

    return { status: StageStatus.SUCCESS };
  },

  rollback: null,
};

File: packages/python/goldenpipe/goldenpipe/adapters/flow.py
"""GoldenFlow adapter -- wraps transform_df()."""
from __future__ import annotations

import logging

from goldenpipe.models.context import PipeContext, StageResult, StageStatus
from goldenpipe.models.stage import StageInfo

logger = logging.getLogger(__name__)

try:
    from goldenflow import transform_df as _transform
    HAS_FLOW = True
except ImportError:
    HAS_FLOW = False
    _transform = None


class TransformStage:
    info = StageInfo(name="goldenflow.transform", produces=["df", "manifest"], consumes=["df"])
    rollback = None

    def validate(self, ctx: PipeContext) -> None:
        if not HAS_FLOW:
            raise RuntimeError("GoldenFlow not installed. Run: pip install goldenpipe[flow]")

    def run(self, ctx: PipeContext) -> StageResult:
        cfg = dict(ctx.stage_config or {})           # COPY: ctx.stage_config IS StageSpec.config
        apply = cfg.pop("apply_repairs", False)       # pop even when False (never leak to transform_df)

        if apply:
            result = self._run_with_repairs(ctx, cfg)
        elif cfg:
            result = _transform(ctx.df, **cfg)
        else:
            result = _transform(ctx.df)

        if hasattr(result, "df"):
            ctx.df = result.df
        if hasattr(result, "manifest"):
            ctx.artifacts["manifest"] = result.manifest
            if "column_contexts" in ctx.artifacts:
                try:
                    from goldenpipe.models.column_context import enrich_contexts_from_flow
                    enrich_contexts_from_flow(ctx.artifacts["column_contexts"], result.manifest)
                except Exception:
                    logger.exception("Failed to enrich column contexts from flow manifest")

        return StageResult(status=StageStatus.SUCCESS)

    def _run_with_repairs(self, ctx: PipeContext, cfg: dict):
        """apply_repairs is on: merge the repair plan's fixer transforms into the
        base GoldenFlowConfig and run explicit mode. Falls through to the normal
        path when there is nothing to apply (keeps auto-detect / byte-identity)."""
        from goldenpipe.repair_host import merge_transforms, repair_transform_specs

        plan = ctx.artifacts.get("repair_plan")
        specs, skipped = repair_transform_specs(plan) if plan else ([], [])
        base = dict(cfg.get("config") or {})
        user_transforms = list(base.get("transforms") or [])

        if not specs and not user_transforms:
            if skipped:
                ctx.reasoning["repair_skipped"] = "; ".join(f"{s['column']}:{s['op']}" for s in skipped)
            # nothing to inject -> behave exactly like the gate-off path
            return _transform(ctx.df, **cfg) if cfg else _transform(ctx.df)

        base["transforms"] = merge_transforms(user_transforms, specs)
        if skipped:
            ctx.reasoning["repair_skipped"] = "; ".join(f"{s['column']}:{s['op']}" for s in skipped)
        logger.info("Applying %d repair transform spec(s); skipped %d assertion(s)", len(specs), len(skipped))
        return _transform(ctx.df, config=base)
